qadence/serialization.py

```diff
@@ -43,27 +47,127 @@ def file_extension(file: Path | str) -> str:
     AbstractBlock,
     QuantumCircuit,
     QuantumModel,
-    QNN,
-    TransformedModule,
     Register,
-    Basic,
+    core.Basic,
     torch.nn.Module,
 ]
 SUPPORTED_TYPES = TypingUnion[
     AbstractBlock,
     QuantumCircuit,
     QuantumModel,
-    QNN,
-    TransformedModule,
     Register,
-    Basic,
+    core.Basic,
     torch.nn.Module,
 ]
 
+EXPR_TYPES = TypingUnion[core.Expr, float, int, str, tuple, dict]
+
 
 ALL_BLOCK_NAMES = [
     n for n in dir(qadenceblocks) if not (n.startswith("__") and n.endswith("__"))
 ] + [n for n in dir(operations) if not (n.startswith("__") and n.endswith("__"))]
+SYMPY_EXPRS = [n for n in dir(core) if not (n.startswith("__") and n.endswith("__"))]
+QADENCE_PARAMETERS = [n for n in dir(parameters) if not (n.startswith("__") and n.endswith("__"))]
+
+
+THIS_PATH = os.path.dirname(__file__)
+GRAMMAR_FILE = os.path.join(THIS_PATH, "serial_expr_grammar.peg")
+
+
+def _parser_fn() -> ParserPEG:
+    with open(GRAMMAR_FILE, "r") as f:
+        grammar = f.read()
+    return ParserPEG(grammar, "Program")
+
+
+def parse_expr_fn(code: str) -> bool:
+    parser = _parser_fn()
+    try:
+        parser.parse(code)
+    except NoMatch as err:
+        print(f"no match error: {err}")
+        return False
+    else:
+        return True
+
+
+@dataclass
+class SerialModel:
```

Comment:
Why is this dataclass and all the child ones needed?
Not implying there would be some easier way to do it, just to understand :)

Reply:
@jpmoutinho  thanks for the questions!

> Are you introducing some new features here that would require also extending the tests?

Not really, as long as the current serialization process works, it's all good, since the new code is made just for that purpose.

> Also, would this make it easier to serialize external models?

Definitely! The `ModelSerial` (line 132) accounts for anything that fits under the "quantum model" category, i.e. inherit either from `QuantumModel` or `torch.nn.Module`. As you can check, I removed the `QNN` and `TransformedModule` from the file imports (and thus from the `SUPPORTED_TYPES`, line 54) and it still works!
 The only catch, and that's something that would eventually happen in the previous case as well, is that the user *needs* to have imported the module (probably with the same name as the serialized code), otherwise it won't work.

> Why is this dataclass and all the child ones needed?

The idea is the dataclasses are "structs" or data holders, so they behave like a data wrapper. The `SerialModel` is there basically to define the attributes every single child will need and also to be easy to generically do this (line 303) (so mypy doesn't complain):

```python
obj: SerialModel
[...]
return obj.value
```

without that, we could not have this generic approach.
Having those dataclasses, makes the process of deserializing much easie, clean and clear (you can check the `deserialize` code (line 262)), because now each dataclass handles the data according to its expected values internally.

Reply:
Indeed! Also, feel free to come up with a better name! I'm not particularly good at naming ^^"